### ISSUE TITLE
some details about MAPE

### ISSUE THREAD

**Issue by wcshen** [StiphyJay/FastPillars#6]
Hello, i'm working on applying the MAPE into my own 3D detector to get better performance on small objects such as cyclist, some details of the MAPE in the arxiv paper confuse me:
here, the attention score  $s_i$ is from $p_i^e$:
![image](https://github.com/StiphyJay/FastPillars/assets/38600406/5a360ae2-6ff0-4e44-b44b-0b022f4fdd5a)

but in the figure2, the attention score  $s_i$ is from $\hat{p_i}$:
![image](https://github.com/StiphyJay/FastPillars/assets/38600406/07685c04-51b7-4ec6-9c36-e30af4c466e6)

**Comment by StiphyJay** [comment]
Thanks for your attention. The attention score is from p_(i)_hat, as shown in Figure 2.  We will update it.
There are some minor mistakes in the arxiv paper. If you have any other questions, please let me know.

**Comment by wcshen** [comment]
I implement the MAPE myself, can you help to check? Especially the dim of the softmax because some details in Fig2 confuses me.
Thanks a lot!
```python
class MAPELayer(nn.Module):
    def __init__(self,
                 in_channels,
                 out_channels):
        super().__init__()
        self.conv_layer1 = nn.Conv2d(in_channels, out_channels, kernel_size=1)
        self.bn1 = nn.BatchNorm2d(out_channels, eps=1e-3, momentum=0.01)
        self.conv_layer2 = nn.Conv2d(in_channels, out_channels, kernel_size=1)
        
    def forward(self, inputs):
        # (1, num_point_features, voxel_count, max_point_per_voxel)
        x = self.conv_layer1(inputs)
        x = self.bn1(x)
        x = F.relu(x)
        # (1, num_point_features, voxel_count, 1)
        max_feature = torch.max(x, dim=-1, keepdim=True)[0]
        # (1, num_point_features, voxel_count, max_point_per_voxel)
        attention_score = F.softmax(self.conv_layer2(inputs), dim=-1)
        avg_feature = x * attention_score
        # (1, num_point_features, voxel_count, 1)
        avg_feature = torch.sum(avg_feature, dim=-1, keepdim=True)
        feature = (avg_feature + max_feature) / 2.0
        return feature
```

**Comment by StiphyJay** [comment]
It seems to be right. Have you tried to add him to your method? Whether there is any performance improvement.

**Comment by wcshen** [comment]
In my experiments on my own dataset, after applied MAPE, the AP of Cyclist increased 7, the AP of Truck dropped 2.

**Comment by StiphyJay** [comment]
Cool! Average performance (map) still has improved, right? If it's useful, welcome to cite this article.

**Comment by ouyangziyao** [comment]
Thank you for your released code about MAPE. The point cloud sampe strategy is dynamic encoding? But the dim of the input of the function contains the point num in pillars and the point num in different pillars is different?  How much is the max num？Additionally, if I plan to use sparse convolution as the backbone, what modifications should I make?

**Comment by wcshen** [comment]
> Thank you for your released code about MAPE. The point cloud sampe strategy is dynamic encoding? But the dim of the input of the function contains the point num in pillars and the point num in different pillars is different? How much is the max num？Additionally, if I plan to use sparse convolution as the backbone, what modifications should I make?

1. In LiDAR 3D object detection, it is a common practice to set the number of points within each pillar to a fixed value. Typically, this involves sampling the original point cloud to ensure that each pillar contains the predetermined number of points. We can call this sample strategy as dynamic encoding, because the sample points result of the same pillar maybe different.
2. you should set the max_points_num per-pillar according to your dataset and pillar-size setting.
3. About the sp-conv, you can reference the code of centerpoint about the backbone, MAPE is just the PFE
